src/hostapi/pulseaudio/pa_linux_pulseaudio.c

```diff
@@ -0,0 +1,1285 @@
+
+
+    /* If this is null we have big problems and we probably are out of memory */
+    if( !c )
+    {
+        PA_PULSEAUDIO_SET_LAST_HOST_ERROR( 0,
+                                           "PaPulseAudio_SourceListCb: Out of memory" );
+        goto error;
+    }
+
+    /* If eol is set to a positive number, you're at the end of the list */
+    if( eol > 0 )
+    {
+        goto error;
+    }
+
+    l_strName = l->name;
+
+    if( l->description != NULL )
+    {
+        l_strName = l->description;
+    }
+
+    if( _PaPulseAudio_AddAudioDevice( l_ptrHostApi,
+                                      l_strName,
+                                      l->name,
+                                      l->sample_spec.channels,
+                                      0,
+                                      (double) l->latency,
+                                      (double) l->configured_latency,
+                                      0,
+                                      0,
+                                      l->sample_spec.rate ) != paNoError )
+    {
+        PA_PULSEAUDIO_SET_LAST_HOST_ERROR( 0,
+                                           "PaPulseAudio_SourceListCb: Can't add device. Maximum amount reached!" );
+    }
+
+  error:
+    pa_threaded_mainloop_signal( l_ptrHostApi->mainloop,
+                                 0 );
+}
+
+/* This routine is called whenever the stream state changes */
+void PaPulseAudio_StreamStateCb( pa_stream * s,
+                                 void *userdata )
+{
+    const pa_buffer_attr *a;
+    /* If you need debug pring enable these
+     * char cmt[PA_CHANNEL_MAP_SNPRINT_MAX], sst[PA_SAMPLE_SPEC_SNPRINT_MAX];
+     */
+
+
+    /* If this is null we have big problems and we probably are out of memory */
+    if( !s )
+    {
+        PA_PULSEAUDIO_SET_LAST_HOST_ERROR( 0,
+                                           "PaPulseAudio_StreamStateCb: Out of memory" );
+        return;
+    }
+
+    switch( pa_stream_get_state(s) )
+    {
+        case PA_STREAM_TERMINATED:
+            break;
+
+        case PA_STREAM_CREATING:
+            break;
+
+        case PA_STREAM_READY:
+             break;
+
+        case PA_STREAM_FAILED:
+        default:
+            PA_DEBUG( ("Portaudio %s: FAILED '%s'\n",
+                      __FUNCTION__,
+                      pa_strerror( pa_context_errno( pa_stream_get_context( s ) ) ) ) );
+
+            break;
+    }
+}
+
+void PaPulseAudio_StreamUnderflowCb( pa_stream * s,
+                                     void *userdata )
+{
+    PaPulseAudio_Stream *stream = (PaPulseAudio_Stream *) userdata;
+
+
+    /* If this is null we have big problems and we probably are out of memory */
+    if( !s )
+    {
+        PA_PULSEAUDIO_SET_LAST_HOST_ERROR( 0,
+                                           "PaPulseAudio_StreamUnderflowCb: Out of memory" );
+        return;
+    }
+
+    stream->underflows++;
+
+    if( stream->underflows >= 6 && stream->latency < 2000000 )
```

Comment:
What does this mean for the using code? Does it need to adjust the buffer?
Is this fix really the business of portaudio?

Reply:
I think this should be removed. It tries to adjust latency if there is underflows but Pulseaudio does it better.

Reply:
If this bounces to bugs then I'll figure out better solution. Now it just create PA_DEBUG and sets last error